text/3417-dropck-eyepatch-v3.md

```diff
@@ -0,0 +1,234 @@
+`#[fully_unused]` or `#[only_dropped]` to that parameter. This **unsafely** asserts
+that the parameter is either completely unused when dropping your type or only
+recursively dropped.
+
+```rust
+struct MyType<'s> {
+    reference: &'s str,
+    needs_drop: String,
+}
+// The impl has to be `unsafe` as the compiler does may not check
+// that `'s` is actually unused.
+unsafe impl<#[only_dropped] 's> Drop for MyType<'s> {
+    fn drop(&mut self) {
+        // println!("{}", reference); // this would be unsound
+        println!("{}", needs_drop);
+    }
+}
+fn can_drop_dead_reference() {
+    let _x;
+    {
+        let temp = String::from("I am only temporary");
+        _x = MyType {
+            reference: &temp,
+            needs_drop: String::from("I have to get dropped"),
+        };
+    }
+    // We drop `_x` here even though `reference` is no longer live.
+    //
+    // This is accepted as `'s` is marked as `#[only_dropped]` in the
+    // `Drop` impl of `MyType`.
+}
+```
+
+The ability to differentiate between `#[fully_unused]` and `#[only_dropped]` is significant
+for type parameters:
+
+```rust
+pub struct BTreeMap<K, V> {
+    root: Option<Root<K, V>>,
+    length: usize,
+}
+
+unsafe impl<#[only_dropped] K, #[only_dropped] V> Drop for BTreeMap<K, V> {
+    fn drop(&mut self) {
+        // Recursively drops the key-value pairs but doesn't otherwise
+        // inspect them, so we can use `#[only_dropped]` here.
+        drop(unsafe {ptr::read(self) }.into_iter())
+    }
+}
+```
+
+A type where `#[fully_unused]` would be useful is a `Weak` pointer for a variant of `Rc`
+where the value is dropped when the last `Rc` goes out of scope. Dropping a `Weak` pointer
+would never access `T` in this case.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Whenever we use a value of a given type this type has to be **well-formed**, requiring
+that all lifetimes in this type are live. An exception to this is the implicit drop when
+a variable goes out of scope. While borrowck ensures that dropping the variable is safe,
+this does not necessarily require all lifetimes to be live.
+
+When implicitly dropping a variable of type `T`, liveness requirements are computed as follows:
+- If `T` does not have any drop glue, do not add any requirements.
+- If `T` is a trait object, `T` has to be live.
+- If `T` has an explicit `Drop` impl, require all generic argument to be live, unless
+    - they are marked with `#[fully_unused]`, in which case they are ignored,
+    - or they are marked with `#[only_dropped]`, in which case recurse into the generic argument.
+- Regardless of whether `T` implements `Drop`, recurse into all types *owned* by `T`:
+    - references, raw pointers, function pointers, function items and scalars do not own
+      anything. They can be trivially dropped.
+    - tuples and arrays consider their element types to be owned.
```

Comment:
for arrays this is currently also a bit inconsistent, see https://github.com/rust-lang/rust/issues/110288 and https://rust-lang.zulipchat.com/#narrow/stream/326866-t-types.2Fnominated/topic/.23110288.3A.20.60.5BT.3B.200.5D.60.20adding.20outlives.20requirements.20to.20dropck for details about this.

Shouldn't matter for this RFC apart from deciding whether to add "`[T; 0]` does not consider `T` to be owned".